src/execution.jl

```diff
@@ -299,26 +327,39 @@ function execute(dsl_file::String,
                  sub_comm_size::Int=1,
                  max_amplitudes::Union{Int, Nothing}=nothing,
                  max_parameters::Union{Int, Nothing}=nothing)
-
+    # Parse the dsl file to create a list of commands to execute in a context. Also parse
+    # the parameter file to get partition parameters and to create a sampler object to 
+    # produce bitstrings to compute amplitudes for.
     commands = parse_dsl(dsl_file)
+    sampler_args, partition_params = parse_parameters(param_file,
+                                                      max_amplitudes=max_amplitudes, 
+                                                      max_parameters=max_parameters)
 
-    bitstrings, partition_params = parse_parameters(param_file,
-                                                    max_amplitudes=max_amplitudes, max_parameters=max_parameters)
-
+    # Create a context to execute the commands in.
     ctx = QXContext(commands, partition_params, input_file)
     if comm !== nothing
         ctx = QXMPIContext(ctx, comm, sub_comm_size)
+
+        # # Variables needed by a sampler to divide work evenly amongst groups of nodes.
+        # sampler_args[:params][:rank] = MPI.Comm_rank(comm) ÷ sub_comm_size
+        # sampler_args[:params][:comm_size] = MPI.Comm_size(comm) ÷ sub_comm_size
     end
 
-    bitstring_iter = BitstringIterator(ctx, bitstrings)
-    results = Array{ComplexF32, 1}(undef, length(bitstring_iter))
-    for (i, bitstring) in enumerate(bitstring_iter)
-        results[i] = compute_amplitude!(ctx, bitstring)
+    # Create a sampler to produce bitstrings to get amplitudes for and a variable to store 
+    # the results.
+    sampler = QXContexts.Sampling.create_sampler(sampler_args, ctx)
```

Comment:
I couldn't think of a way to organise the changes to avoid calling `QXContexts.Sampling`.

Reply:
I think the problem here is that QXContext should not be defined in the Execution module, but should probably instead be in its own file/module. I think as it is can work for now and when it comes to adding context for the GPU we can look at splitting it out.